Fixes #11085 - Email reporting for openscap policies

xprazak2 opened the pull request:
User can subscribe to receive a regular summary of OpenScap policies. Summary contains overview of policies with host status and if there has been any recent change in host compliance.

shlomizadok commented on app/models/concerns/foreman_openscap/host_extensions.rb:
```diff
@@ -30,6 +30,16 @@ def combined_policies
+      !last_reports.first.eql? last_reports.last
+    end
+
+    def reports_for_policy(policy)
```
Not needed. Use: policy.arf_reports

xprazak2 replied:
I need reports for host that belong to a certain policy: `policy.arf_reports.select { |report| report.asset = self.asset }`

I made a slight improvement by adding `limit`

shlomizadok commented on db/seeds.d/openscap_policy_notification.rb:
```diff
@@ -0,0 +1,9 @@
+  :subscription_type => 'report',
+}
+
+MailNotification.create(policy_notification) if MailNotification.where(:name => policy_notification[:name]).blank?
```
Maybe use ```first_or_create``` ?

shlomizadok replied:
nitpick: extra line

shlomizadok commented on test/unit/policy_mailer_test.rb:
```diff
@@ -0,0 +1,38 @@
+  test 'policy mailer sends Foreman URL in body' do
+    assert @email.body.include? Setting[:foreman_url]
+  end
+end
```
nitpick: extra line

xprazak2 replied:
The line 39 should be there, otherwise git would show the no-newline-at-the-end-of-the-file symbol. Same goes for the following files.

shlomizadok replied:
That was my meaning too... :)

shlomizadok commented on test/factories/xccdf_rule_result_factory.rb:
```diff
@@ -0,0 +1,6 @@
+    xccdf_result_id 1
+    xccdf_rule_id 1
+  end
+end
```
nitpick: extra line

shlomizadok commented on test/factories/arf_report_factory.rb:
```diff
@@ -4,5 +4,7 @@
+    xccdf_rule_results []
   end
+
 end
```
nitpick (to myself...): extra line

shlomizadok commented on app/views/foreman_openscap/policy_mailer/_hosts.erb:
```diff
@@ -0,0 +1,44 @@
+      </td>
+    <% else %>
+      <td style="<%= td_style %> border-right: 0px;"> 
+        <%= _('No reports for this policy') %>
```
Text: No ARF reports for this policy

shlomizadok commented on app/models/concerns/foreman_openscap/policy_extensions.rb:
```diff
@@ -143,6 +155,10 @@ def to_enc
       }.merge(period_enc)
     end
 
+    def find_assets(method)
```
Isn't this similar to <Policy>.assets?

xprazak2 replied:
not quite, I use it to call scopes on assets comfortably, `policy.find_assets(:incomply_with)` is really `Scaptimony::Asset.incomply_with(policy)`

shlomizadok replied:
To be honest, I'm a little uncomfortable with this method, as it can send anything to ```Scaptimony::Asset```.
I think that something like: ```policy.assets.incomply_with(policy)``` || ```policy.assets.comply_with(policy)```  would be sufficient for this use case. @ares - would love to hear your comments on this.

ares replied:
definitely a `public_send` would be better than `send` but I tend to agree with @shlomizadok, I found just two usages of this method for so I'd prefer to be explicit, something like

```ruby
def find_assets(scope)
  case scope
  when :compliant
    Scaptimony::Asset.comply_with(self)
  when :incompliant
    Scaptimony::Asset.incomply_with(self)
  else
    raise "Unsupported asset scope: #{scope}"
end
```

shlomizadok commented on app/mailers/foreman_openscap/policy_mailer.rb:
```diff
@@ -0,0 +1,47 @@
+
+    def hosts_of_policies(policies, method)
+      assets = policies.map { |policy| policy.find_assets(method)}.flatten
+      Host.where(:id => assets.map(&:assetable_id))
```
```Host.find(assets.hosts.map(&:assetable_id))``` ?

shlomizadok commented on app/mailers/foreman_openscap/policy_mailer.rb:
```diff
@@ -0,0 +1,47 @@
+    private
+
+    def hosts_of_policies(policies, method)
+      assets = policies.map { |policy| policy.find_assets(method)}.flatten
```
```policy.assets``` ?

xprazak2 replied:
same as above, I need lists of compliant and incompliant assets

shlomizadok commented on app/models/concerns/foreman_openscap/host_extensions.rb:
```diff
@@ -30,6 +30,16 @@ def combined_policies
     end
 
+    def scap_status_changed?(policy)
+      last_reports = reports_for_policy(policy).take 2
```
maybe last two should be in the policy's scope? something like ```Scaptimony::Policy.find(:all, :order => "id desc", :limit => 5)``` ?

shlomizadok commented on app/models/concerns/foreman_openscap/arf_report_extensions.rb:
```diff
@@ -46,5 +46,13 @@ def failed?
       passed > 0 && !failed?
     end
+
+    def eql?(other)
```
Method name should be ```equal?```

shlomizadok commented on app/views/foreman_openscap/policy_mailer/_dashboard.erb:
```diff
@@ -0,0 +1,24 @@
+  <tbody>
+    <tr>
+      <td style="font-size: 125%; width: 33%; padding: 5px; border: 1px solid #ccc;">
+        <%= (n_('<b style="color: #6bb5df;">%{changed_hosts}</b>
```
html style should not be extracted into translation

shlomizadok commented on app/views/foreman_openscap/policy_mailer/_dashboard.erb:
```diff
@@ -0,0 +1,24 @@
+                '<b style="color: #6bb5df;">%{changed_hosts}</b>
+                <p style="font-weight: bold; margin: 5px; padding: 0px;">Changed</p>', changed_hosts) % {:changed_hosts => changed_hosts.length}).html_safe %>
+      </td>
+      <td style="font-size: 125%; width: 33%; padding: 5px; border: 1px solid #ccc;">
```
Instead of repeating the html style use it in a variable.
See [style in variable](https://github.com/theforeman/foreman/blob/develop/app/views/host_mailer/_active_hosts.html.erb#L1) and [usage](https://github.com/theforeman/foreman/blob/develop/app/views/host_mailer/_active_hosts.html.erb#L4)

shlomizadok commented on app/views/foreman_openscap/policy_mailer/_list.erb:
```diff
@@ -0,0 +1,10 @@
+  <% else %>
+    <b><%= _('None!') %></b>
+  <% end %>
+</div>
```
nitpick: extra line

shlomizadok commented:
I love this PR :+1: 
Got some comments to fix.

xprazak2 commented:
I made changes according to comments, but I left the the `find_assets` method in policy_extension and `reports_for_policy` in host_extension for now because I think I need them. Reasons are explained above.

xprazak2 commented:
I removed the extra line at the end of files - not sure why I thought they belonged there. I also removed the  `find_assets` method from policy_extension - I guess I was looking for needlessly complex solution when there was a simpler one available :-)

shlomizadok commented:
@xprazak2 -- I am so sorry for the confusion, we actually _do_ need the extra line at the end. :confused:
Mind adding them back? Thanks a lot

xprazak2 commented:
the new lines are back! :-D

shlomizadok commented on db/seeds.d/openscap_policy_notification.rb:
```diff
@@ -0,0 +1,9 @@
+  :subscription_type => 'report',
+}
+
+MailNotification.first_or_create(policy_notification)
```
This is a wrong syntax, the mail notification will not be created.
The syntax should be:
```ruby
MailNotification.where(policy_notification).first_or_create
```

shlomizadok commented on app/mailers/foreman_openscap/policy_mailer.rb:
```diff
@@ -0,0 +1,41 @@
+
+    def changed_hosts_of_policies(policies)
+      hash = @policies.inject({}) do |result, policy|
+        result[policy.id] = policy.hosts
```
Please note that this will fail if the host is deleted. See this [workaround](https://github.com/theforeman/foreman_openscap/blob/f4a4d624a76e10838abbc3714d2cdd014b0276be/app/helpers/concerns/foreman_openscap/hosts_helper_extensions.rb#L46) to solve this

shlomizadok commented:
@xprazak2 found two issues while testing.

xprazak2 commented:
I fixed the seed file and deleting a host no longer results in error. If we added `:dependent => :nullify` [here](https://github.com/theforeman/foreman_openscap/blob/master/app/models/concerns/foreman_openscap/host_extensions.rb#L9), wouldn't it make our lives a little easier?

shlomizadok commented:
Works like a charm.
